test/schema.validation.test.js

```diff
@@ -1330,6 +1330,26 @@ describe('schema', function() {
       });
     });
 
+    it('Allows for doc to be passed as another parameter gh-12564', function(done) {
+      const s = mongoose.Schema({
+        n: {
+          type: String,
+          // required: true,
+          required: [true, function(properties, doc) {
```

Comment:
Use a custom validator instead of `required`, so something like:

```
validate: {
  validator: function(v) { ... },
  message: function(properties, doc) { ... }
}
```